src/builder/V2DutchOrderBuilder.ts

```diff
@@ -0,0 +1,246 @@
+      super.deadline(decayEndTime);
+    }
+
+    this.info.cosignerData.decayEndTime = decayEndTime;
+    return this;
+  }
+
+  input(input: DutchInput): this {
+    this.info.input = input;
+    return this;
+  }
+
+  output(output: DutchOutput): this {
+    invariant(
+      output.startAmount.gte(output.endAmount),
+      `startAmount must be greater than endAmount: ${output.startAmount.toString()}`
+    );
+    this.info.outputs.push(output);
+    return this;
+  }
+
+  deadline(deadline: number): this {
+    super.deadline(deadline);
+    if (!this.info.cosignerData.decayEndTime) {
+      this.decayEndTime(deadline);
+    }
+    return this;
+  }
+
+  swapper(swapper: string): this {
+    super.swapper(swapper);
+    return this;
+  }
+
+  nonce(nonce: BigNumber): this {
+    super.nonce(nonce);
+    return this;
+  }
+
+  validation(info: ValidationInfo): this {
+    super.validation(info);
+    return this;
+  }
+
+  // ensures that we only change non fee outputs
+  nonFeeRecipient(newRecipient: string, feeRecipient?: string): this {
+    invariant(
+      newRecipient !== feeRecipient,
+      `newRecipient must be different from feeRecipient: ${newRecipient}`
+    );
+    if (!this.info.outputs) {
+      return this;
+    }
+    this.info.outputs = this.info.outputs.map((output) => {
+      // if fee output then pass through
+      if (
+        feeRecipient &&
+        output.recipient.toLowerCase() === feeRecipient.toLowerCase()
+      ) {
+        return output;
+      }
+
+      return {
+        ...output,
+        recipient: newRecipient,
+      };
+    });
+    return this;
+  }
+
+  exclusiveFiller(exclusiveFiller: string): this {
+    this.info.cosignerData.exclusiveFiller = exclusiveFiller;
+    return this;
+  }
+
+  inputOverride(inputOverride: BigNumber): this {
+    this.info.cosignerData.inputOverride = inputOverride;
+    return this;
+  }
+
+  outputOverrides(outputOverrides: BigNumber[]): this {
+    this.info.cosignerData.outputOverrides = outputOverrides;
+    return this;
+  }
+
+  cosigner(cosigner: string): this {
+    this.info.cosigner = cosigner;
+    return this;
+  }
+
+  cosignature(cosignature: string | undefined): this {
+    this.info.cosignature = cosignature;
+    return this;
+  }
+
+  cosignerData(cosignerData: CosignerData): this {
+    this.decayStartTime(cosignerData.decayStartTime);
+    this.decayEndTime(cosignerData.decayEndTime);
+    this.exclusiveFiller(cosignerData.exclusiveFiller);
+    this.inputOverride(cosignerData.inputOverride);
+    this.outputOverrides(cosignerData.outputOverrides);
+    return this;
+  }
+
+  build(): V2DutchOrder {
+    invariant(this.info.cosigner !== undefined, "cosigner not set");
+    invariant(this.info.input !== undefined, "input not set");
+    invariant(this.info.outputs.length > 0, "outputs not set");
+    invariant(
+      this.info.cosignerData.decayStartTime !== undefined,
+      "decayStartTime not set"
+    );
+    invariant(
```

Comment:
might want to allow building without cosigner data for the pre-auction signing flow